jubilant/_juju.py

```diff
@@ -456,6 +456,27 @@ def integrate(self, app1: str, app2: str, *, via: str | Iterable[str] | None = N
                 args.extend(['--via', ','.join(via)])
         self.cli(*args)
 
+    def offer(self, app: str, endpoint: str | Iterable[str], *, name: str | None = None) -> None:
+        """Offer application endpoints for use in other models.
+
+        Args:
+            app: Application name to offer endpoints for.
+            endpoint: Endpoint or endpoints to offer, for example ``'db'`` or ``['db', 'log']``.
+            name: Name of the offer. By default, the offer is named after the application.
+        """
+        if ':' in app:
+            raise TypeError('must provide endpoint in "endpoint" argument not in "app"')
```

Comment:
Is the departure from CLI shorthand intentional?
I had just done `juju offer tempo:tracing` yesterday.

Reply:
Yeah, but I'm unsure about it -- see the PR description. I think I might switch it back, especially given your feedback. F-strings are so easy to use these days if you do have them as separate variables: `f'{app}:{endpoint}'` (though it seems much more likely it'll be hard-coded).

Reply:
My thoughts are pretty identical to @james-garner-canonical 's. A "app:e1,e2,e3" string feels very raw and not in sync with the Jubilant goals. It's easy to combine strings, but it's also easy to split them, so I don't find that compelling either way. `juju.offer('tempo', 'tracing')` seems a very reasonable interpretation to me, not a departure.

I like the `*args` suggestion @james-garner-canonical made, particularly given the use in other places.

It is inconsistent with integrate, but it's different anyway since you can only have a single endpoint per app in that case. It's also less common to need to specify an endpoint there, whereas you *have* to have one here.

Reply:
Cool, I've gone with `*endpoint`, so `juju.offer('tempo', 'tracing')`.